### PR TITLE
HPCC-28158 replace expired EFS/CSI provisioner with EFS CSI driver

### PR DESCRIPTION
**Pull request by xwang2713**
Previous EFS/CSI provisioner is no longer available so the examples/efs is broken. This fix follows AWS documentation https://docs.aws.amazon.com/eks/latest/userguide/efs-csi.html.

Current implementation assume a IAM user is created with proper permissions for EKS, Open ID Connect Provider, etc.
I contacted RISK AWS support for ADFS and still wait their reply (There is already a ticket in Open ID Connect Provider). 
We don't want to wait ADFS issue resolved because there is demo for Customer 2 on AWS. Also there is no working storage example on AWS now for community users.

Users should follow examples/efs/README.md to install the EFS CSI provisioner. The basic steps are:
1) Fill AWS Account, EKS, EFS, etc in efs-env
2) run install-csi-driver.sh
3) Create Storage Class: kubectl apply -f storageclass.yaml
4) Deploy HPCC AWS storage (skip  if want to use auto storage)
5) Deploy HPCC cluster with the probably storage yaml file

@g-pan please view examples/efs/README.md which replaces examples/efs/hpcc-efs/README.md And it is very similar to README in https://github.com/hpcc-systems/HPCC-For-AWS

Signed-off-by: Xiaoming Wang <xiaoming.wang@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:

Tested in AWS EKS cluster.  Tested both auto storage and installing pv/pvc before deploy HPCC

### CODE REVIEW

File: helm/examples/efs/README.md [hunk 1]
```diff
@@ -1,51 +1,69 @@
...
-The output will display EFS name and ID.
 
-It is recommended to provide the option "--managed" and a security group id during EKS cluster creation. EFS Server can use the same security group or if you know your security group of node pools they are OK to use for EFS security group. Otherwise you need to provide the following:
-- EKS cluster name
-```console
-kubectl config get-clusters | cut -d'.' -f1
+
+## EFS CSI Driver
+"examples/efs" requires Elastic FileSystem (EFS) Container Storage Interface (CSI) driver.
+We provide some scripts to simplified the process. For detail please referencing https://docs.aws.amazon.com/eks/latest/userguide/efs-csi.html
```
Comment (jakesmith) [inline helm/examples/efs/README.md:7]: simplified -> simplify

Comment (jakesmith) [reply helm/examples/efs/README.md:7]: For detail please referencing  -> For details please refer to:

File: helm/examples/efs/README.md [hunk 2]
```diff
@@ -1,51 +1,69 @@
...
+### Prepare environment
+Make sure an EFS server is available or create an [EFS server](#efs-server) before using this chart.<br/>
+Fill following information in efs-env
+```code
+ACCOUNT_ID       # AWS Account ID
+EKS_NAME         # Elastic Kubernetes Service (EKS) cluster name: kubectl configure get-cluster  (The first part of the name before ".")
+EFS_ID           # EFS ID can be found from AWS Console or command-line:
+                   aws efs describe-file-systems --query "FileSystems[*].FileSystemId" --output text
+EFS_REGION       # EFS region. We strongly suggest you have the same region for EFS server and EKS cluster
+EFS_SECURITY_GROUPS  # EFS seucurity group which is the subnet security group which should be same as EKS
```
Comment (jakesmith) [inline helm/examples/efs/README.md:19]: seucurity -> security

File: helm/examples/efs/README.md [hunk 3]
```diff
@@ -1,51 +1,69 @@
...
+EFS_BASE_PATH    # Base directory to mount in the EFS. If you use EFS for multiple HPCC clusters you should set this differently to avoid overwriting each others
```
Comment (jakesmith) [inline helm/examples/efs/README.md:20]: to avoid overwriting each others -> to avoid clashing with the other clusters

File: helm/examples/efs/README.md [hunk 4]
```diff
@@ -96,8 +108,7 @@ Pick a VPC in the the same region of EFS server
   aws ec2 describe-vpcs --region <REGION>
 ```
 The output shows the VPC IDs.
-
-Get all the subnets of the VPC
+ll the subnets of the VPC
```
Comment (jakesmith) [inline helm/examples/efs/README.md:111]: > ll the subnets of the VPC

I assume this is supposed to still say "Get all the subnets of the VPC" ?

File: helm/examples/efs/associate-oidc.sh [hunk 5]
```diff
@@ -0,0 +1,18 @@
+#!/bin/bash
+WORK_DIR=$(dirname $0)
+
+source ${WORK_DIR}/efs-env
+
+oidc_id=$(aws eks describe-cluster --name ${EKS_NAME} --query "cluster.identity.oidc.issuer" --output text | cut -d '/' -f 5)
+echo "oidc_id: $oidc_id"
+oidc_providers=$(aws iam list-open-id-connect-providers | grep $oidc_id)
+if [ "${oidc_providers}" = "" ]
```
Comment (jakesmith) [inline helm/examples/efs/associate-oidc.sh:9]: trivial: more common to test empty with -z :
```
if [ -z "${oidc_providers}" ]
```

**Review by jakesmith — COMMENTED** [review]
@xwang2713 - looks ok afaics, some minor comments.

File: helm/examples/efs/README.md (hunk 1 above)
Comment (g-pan) [inline helm/examples/efs/README.md:7]: s/b:  We provide some scripts to simplify the process. For details please reference

File: helm/examples/efs/README.md [hunk 6]
```diff
@@ -1,51 +1,69 @@
...
-kubectl config get-clusters | cut -d'.' -f1
+
+## EFS CSI Driver
+"examples/efs" requires Elastic FileSystem (EFS) Container Storage Interface (CSI) driver.
+We provide some scripts to simplified the process. For detail please referencing https://docs.aws.amazon.com/eks/latest/userguide/efs-csi.html
+Current implementation only tested successfully with a IAM user. It may need update for Active Directory Federation Service (ADFS)
+
+### Prepare environment
+Make sure an EFS server is available or create an [EFS server](#efs-server) before using this chart.<br/>
+Fill following information in efs-env
```
Comment (g-pan) [inline helm/examples/efs/README.md:12]: Provide the following information in efs-env

File: helm/examples/efs/README.md [hunk 7]
```diff
@@ -1,51 +1,69 @@
...
-## Deploy efs-provisioner
-  efs-provisioner pod should be started first:
-```console
-  ./install-efs-provision.sh
-  # To check the pod:
-  kubectl get pod
-```
-There may be some warnings which can be ignored:
+### Install EFS CSI Driver
+Run following command:
```
Comment (g-pan) [inline helm/examples/efs/README.md:38]: Run the following command:

File: helm/examples/efs/README.md (hunk 4 above)
Comment (g-pan) [reply helm/examples/efs/README.md:111]: Get all the subnets of the VPC

**Review by g-pan — COMMENTED** [review]
see comments in line

File: helm/examples/efs/README.md (hunk 1 above)
Comment (xwang2713) [reply helm/examples/efs/README.md:7]: Fixed

File: helm/examples/efs/README.md (hunk 2 above)
Comment (xwang2713) [reply helm/examples/efs/README.md:19]: fixed

File: helm/examples/efs/README.md (hunk 3 above)
Comment (xwang2713) [reply helm/examples/efs/README.md:20]: fixed

File: helm/examples/efs/README.md (hunk 4 above)
Comment (xwang2713) [reply helm/examples/efs/README.md:111]: Fixed

File: helm/examples/efs/associate-oidc.sh (hunk 5 above)
Comment (xwang2713) [reply helm/examples/efs/associate-oidc.sh:9]: fixed

File: helm/examples/efs/README.md (hunk 1 above)
Comment (xwang2713) [reply helm/examples/efs/README.md:7]: fixed

File: helm/examples/efs/README.md (hunk 6 above)
Comment (xwang2713) [reply helm/examples/efs/README.md:12]: fixed

File: helm/examples/efs/README.md (hunk 7 above)
Comment (xwang2713) [reply helm/examples/efs/README.md:38]: fixed

**Review by g-pan — APPROVED** [review]
looks good

**Comment by xwang2713** [comment]
@jakesmith can you review it again? If OK I can rebase it. Thanks

File: helm/examples/efs/trust-policy.json.template [hunk 8]
```diff
@@ -0,0 +1,18 @@
+{
+  "Version": "2012-10-17",
```
Comment (jakesmith) [inline helm/examples/efs/trust-policy.json.template:2]: is this template (and helm/examples/efs/iam-policy-example.json) copied from somewhere?
Does it need a reference/acknolegement comment? 
Or is the Version = 2012-10-17 correct , if it is now versioning?

**Review by jakesmith — COMMENTED** [review]
@xwang2713 - 1 question

File: helm/examples/efs/trust-policy.json.template (hunk 8 above)
Comment (xwang2713) [reply helm/examples/efs/trust-policy.json.template:2]: Copied from AWS. It is always " Version = 2012-10-17 "

Comment (jakesmith) [reply helm/examples/efs/trust-policy.json.template:2]: ok

**Review by jakesmith — APPROVED** [review]
@xwang2713 - looks good afaics, please squash

**Comment by jakesmith** [comment]
@xwang2713 - I suspect this should now target 8.10.x ?

**Comment by xwang2713** [comment]
@richardkchapman this is ready to merge.  It will not break anything. Only helm/examples/efs update. Current efs example is not usable anyway (the efs provisioner expired).
